python/ray/data/_internal/datasource/lance_datasink.py

```diff
@@ -0,0 +1,266 @@
+
+if TYPE_CHECKING:
+    import pandas as pd
+    from lance.fragment import FragmentMetadata
+
+
+def _pd_to_arrow(
+    df: Union["pa.Table", "pd.DataFrame", Dict], schema: Optional[pa.Schema]
+) -> "pa.Table":
+    """Convert a pandas DataFrame to pyarrow Table."""
+    import pandas as pd
+
+    if isinstance(df, dict):
+        return pa.Table.from_pydict(df, schema=schema)
+    elif isinstance(df, pd.DataFrame):
+        tbl = pa.Table.from_pandas(df, schema=schema)
+        tbl = tbl.replace_schema_metadata()
+        return tbl
+    elif isinstance(df, pa.Table):
+        if schema is not None:
+            return df.cast(schema)
+    return df
+
+
+def _write_fragment(
+    stream: Iterable[Union["pa.Table", "pd.DataFrame"]],
+    uri: str,
+    *,
+    schema: Optional["pa.Schema"] = None,
+    max_rows_per_file: int = 1024 * 1024,
+    max_bytes_per_file: Optional[int] = None,
+    max_rows_per_group: int = 1024,  # Only useful for v1 writer.
+    data_storage_version: Optional[str] = None,
+    storage_options: Optional[Dict[str, Any]] = None,
+) -> List[Tuple["FragmentMetadata", "pa.Schema"]]:
+    import pandas as pd
+    from lance.fragment import write_fragments, DEFAULT_MAX_BYTES_PER_FILE
+
+    if schema is None:
+        first = next(stream)
+        if isinstance(first, pd.DataFrame):
+            schema = pa.Schema.from_pandas(first).remove_metadata()
+        elif isinstance(first, Dict):
+            tbl = pa.Table.from_pydict(first)
+            schema = tbl.schema.remove_metadata()
+        else:
+            schema = first.schema
+        if len(schema.names) == 0:
+            # Empty table.
+            schema = None
+
+        stream = chain([first], stream)
+
+    def record_batch_converter():
+        for block in stream:
+            tbl = _pd_to_arrow(block, schema)
+            yield from tbl.to_batches()
+
+    max_bytes_per_file = (
+        DEFAULT_MAX_BYTES_PER_FILE if max_bytes_per_file is None else max_bytes_per_file
+    )
+
+    reader = pa.RecordBatchReader.from_batches(schema, record_batch_converter())
+    fragments = write_fragments(
+        reader,
+        uri,
+        schema=schema,
+        max_rows_per_file=max_rows_per_file,
+        max_rows_per_group=max_rows_per_group,
+        max_bytes_per_file=max_bytes_per_file,
+        data_storage_version=data_storage_version,
+        storage_options=storage_options,
+    )
+    return [(fragment, schema) for fragment in fragments]
+
+
+class _BaseLanceDatasink(Datasink):
+    """Base class for Lance Datasink."""
+
+    def __init__(
+        self,
+        uri: str,
+        schema: Optional[pa.Schema] = None,
+        mode: Literal["create", "append", "overwrite"] = "create",
+        storage_options: Optional[Dict[str, Any]] = None,
+        *args,
+        **kwargs,
+    ):
+        super().__init__(*args, **kwargs)
+
+        self.uri = uri
+        self.schema = schema
+        self.mode = mode
+
+        self.read_version: int | None = None
```

Comment:
```suggestion
        self.read_version: Optional[int] = None
```
Only Python 3.10+ supports `int | None`

Reply:
Fixed.